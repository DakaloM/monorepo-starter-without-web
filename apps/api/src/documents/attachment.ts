import { BaseModel } from '@repo/datakit';
import mime from 'mime-types';
import path from 'path';
import { User } from '~/account/user';
import { Address } from '~/address/address';
import { Assignee } from '~/assignee/assignee';
import { Validator } from '~/domain/validation';
import { Event } from '~/event';
import { Type, TypeIdentifier } from '~/type/type';

import { ContentType, contentTypes } from './contentTypes';

export { AttachmentVersion } from './version';

export enum AttachmentState {
  IN_PROGRESS = 'IN_PROGRESS',
  COMPLETED = 'COMPLETED',
}

export const maxContentLength = 100 * 1024 * 1024; // 100mb
export class Attachment extends BaseModel {
  assignees: Assignee[];
  contentType: ContentType;
  sequence: number;
  contentLength!: number;
  creator: User;
  creatorId: string;
  deletedAt: Date | null;
  description: string;
  id: string;
  confidential: boolean;
  createdAt!: Date;
  name: string;
  refId: string;
  state: AttachmentState;
  refTypeId: string;
  versionCount: number;
  updatedAt: Date;
  events: Event<Attachment>[];
  address: Address | null;

  static tableName = 'attachment';
  static typeIdentifier = TypeIdentifier.Attachment;

  static createSchema() {
    return Validator.create((z, is) =>
      z.object({
        contentType: z.enum(contentTypes),
        contentLength: z.number().positive().max(maxContentLength),
        description: z.string().trim().min(1).max(5000).optional(),
        creatorId: z.string().uuid(),
        name: z.string().trim().superRefine(is.filename),
        refId: z.string().uuid(),
        refTypeId: z.string().uuid(),
        state: z.nativeEnum(AttachmentState),
        confidential: z.boolean(),
        sequence: z.number().positive().min(1),
      }),
    );
  }

  static updateSchema() {
    return Validator.create((z, is) =>
      z.object({
        description: z.string().trim().min(1).max(5000).optional(),
        editorId: z.string().uuid(),
        name: z.string().trim().min(1).max(5000).optional(),
        state: z.nativeEnum(AttachmentState).optional(),
        updatedAt: z.date(),
      }),
    );
  }

  static parseContentType(filename: string) {
    const ext = path.extname(filename);

    const mimeType = mime.contentType(ext) as string;

    return mimeType;
  }

  static get relationMappings() {
    return {
      assignees: {
        relation: BaseModel.HasManyRelation,
        modelClass: Assignee,
        join: {
          from: 'attachment.id',
          to: 'assignee.refId',
        },
      },
      address: {
        relation: BaseModel.BelongsToOneRelation,
        modelClass: Address,
        join: {
          from: 'attachment.fileId',
          to: 'address.refId',
        },
      },
      creator: {
        relation: BaseModel.BelongsToOneRelation,
        modelClass: User,
        join: {
          from: 'attachment.creatorId',
          to: 'user.id',
        },
      },
      refType: {
        relation: BaseModel.BelongsToOneRelation,
        modelClass: Type,
        join: {
          from: 'attachment.refTypeId',
          to: 'type.id',
        },
      },

      events: {
        relation: BaseModel.HasManyRelation,
        modelClass: Event,
        join: {
          from: 'attachment.id',
          to: 'event.refId',
        },
      },
    };
  }
}
