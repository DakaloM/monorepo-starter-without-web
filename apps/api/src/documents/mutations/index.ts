export { createAttachment } from './createAttachment';
export { trashAttachment } from './trashAttachment';
