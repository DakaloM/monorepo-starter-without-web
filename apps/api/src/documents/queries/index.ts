export { getAttachmentById } from './getAttachmentById';

export { getAttachmentsByRef } from './getAttachmentsByRef';

export { getAttachments } from './getAttachments';
