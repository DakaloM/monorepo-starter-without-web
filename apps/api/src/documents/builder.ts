import { createBuilder } from '@num/datakit';
import { faker } from '@num/testkit';

import { Context } from '~/context';
import { TypeIdentifier } from '~/type';

import { Attachment, AttachmentState } from './attachment';

export const attachmentBuilder = createBuilder(
  // @ts-ignore
  async (attrs: Partial<Attachment>, _factory, ctx: Context) => {
    const refId = attrs.refId as string;
    const sequence = await ctx.bumpSequenceValue(refId, TypeIdentifier.Attachment);
    return Attachment.fromJson({
      name: faker.system.commonFileName('pdf'),
      contentType: 'application/pdf',
      contentLength: 100,
      state: AttachmentState.COMPLETED,
      sequence,
      ...attrs,
    });
  },
);
