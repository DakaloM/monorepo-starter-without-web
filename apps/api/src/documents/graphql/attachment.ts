import { builder } from '~/graphql/builder';

import { allow } from 'graphql-shield';
import { isActiveUser } from '~/account/shield';
import { AddressSchema } from '~/address/graphql';
import { HasAssigneesType } from '~/assignee/graphql';
import { getAttachmentById, getAttachments } from '~/documents/queries';
import { HasEventsSchema } from '~/event/graphql/event';

import { Attachment, AttachmentState, AttachmentVersion } from '../attachment';

export const AttachmentStateType = builder.enumType(AttachmentState, {
  name: 'AttachmentState',
  description: 'Attachment state',
});

export const AttachmentSchema = builder.objectType(Attachment, {
  name: 'Attachment',
  description: 'An attachment to an object.',
  interfaces: [HasAssigneesType, HasEventsSchema],
  shield: isActiveUser,
  fields: (t) => ({
    id: t.exposeID('id'),
    contentLength: t.exposeInt('contentLength', {
      description: 'The size of the attachment in bytes.',
      nullable: false,
    }),
    refId: t.exposeString('refId', {}),
    sequence: t.exposeInt('sequence', {}),
    contentType: t.exposeString('contentType', {
      description: 'The content type of the attachment.',
      nullable: false,
    }),
    description: t.exposeString('description', {
      description: 'The description rendered to text of the attachment.',
      nullable: true,
    }),
    name: t.exposeString('name', {
      description: 'The name of the attachment.',
      nullable: false,
    }),
    state: t.expose('state', {
      type: AttachmentStateType,
      description: 'The state of the attachment.',
    }),
    address: t.expose('address', {
      type: AddressSchema,
      description: 'The address of the attachment.',
      nullable: true,
    }),
    versions: t.field({
      type: [AttachmentVersionType],
      description: 'Versions of the attachment.',
      resolve: async (parent, _args, ctx) => {
        const versions = await ctx.storageClient.getVersions(parent.id);

        return versions.map(
          (version) =>
            new AttachmentVersion(
              parent.id,
              parent.name,
              version.id,
              version.id,
              version.createdAt,
            ),
        );
      },
    }),
    uploadUrl: t.field({
      type: 'String',
      description: 'The URL to upload the attachment.',
      resolve: async (parent, args, ctx) => {
        const url = await ctx.storageClient.getUploadURL({
          path: parent.id,
          contentType: parent.contentType,
          contentLength: parent.contentLength,
        });

        return url;
      },
    }),
    downloadUrl: t.field({
      type: 'String',
      description: 'The URL to download the attachment.',
      resolve: (parent, _args, ctx) =>
        ctx.storageClient.getDownloadURL({
          name: parent.name,
          path: parent.id,
        }),
    }),
    createdAt: t.expose('createdAt', {
      type: 'Date',
      description: 'The creation date of the case',
    }),
  }),
});

export const AttachmentVersionType = builder.objectType(AttachmentVersion, {
  name: 'AttachmentVersion',
  description: 'A version of an attachment.',
  fields: (t) => ({
    attachmentId: t.exposeString('attachmentId', {
      description: 'The ID of the attachment.',
      nullable: false,
    }),
    attachmentName: t.exposeString('attachmentName', {
      description: 'The name of the attachment.',
      nullable: false,
    }),
    versionId: t.exposeString('versionId', {
      description: 'The ID of the version.',
      nullable: true,
    }),
    version: t.exposeString('version', {
      description: 'The version of the attachment.',
      nullable: false,
    }),
    insertedAt: t.field({
      type: 'Date',
      description: `The date and time when the version was created.`,
      nullable: true,
      resolve: (parent) => parent.insertedAt,
    }),
    downloadUrl: t.field({
      type: 'String',
      description: 'The URL to download the attachment.',
      resolve: (parent, _args, ctx) =>
        ctx.storageClient.getVersionDownloadURL(
          { name: parent.attachmentName, path: parent.attachmentId },
          parent.versionId,
        ),
    }),
  }),
});

builder.queryField('attachment', (t) =>
  t.field({
    type: AttachmentSchema,
    shield: allow,
    args: {
      id: t.arg.id({ required: true }),
    },
    resolve: async (_parent, args, ctx) => getAttachmentById(args.id.toString(), ctx),
  }),
);
const AttachmentFilter = builder.inputType('AttachmentFilter', {
  fields: (t) => ({
    limit: t.field({ type: 'Int', defaultValue: 20 }),
    page: t.field({ type: 'Int', defaultValue: 1 }),
  }),
});

builder.queryField('attachments', (t) =>
  t.field({
    type: [AttachmentSchema],
    shield: allow,
    args: {
      input: t.arg({
        type: AttachmentFilter,
        required: true,
      }),
    },
    resolve: async (_parent, args, ctx) =>
      getAttachments(ctx, {
        limit: args.input.limit || undefined,
        page: args.input.page || undefined,
      }),
  }),
);
