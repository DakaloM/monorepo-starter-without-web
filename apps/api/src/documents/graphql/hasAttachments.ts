import { BaseModelInstance } from '@num/datakit';

import { builder } from '~/graphql/builder';

import { getAttachmentsByRef } from '~/documents/queries';

import { AttachmentSchema } from './attachment';

export const HasAttachmentsType = builder.interfaceType('HasAttachments', {
  fields: (t) => ({
    attachments: t.field({
      type: [AttachmentSchema],
      args: {},
      description: 'The list of the attachments of the object.',
      resolve: async (root, _args, ctx) => {
        const parent = root as BaseModelInstance;
        const type = await ctx.resolveTypeAssert(parent);

        const result = await getAttachmentsByRef(
          {
            refId: parent.id,
            refTypeId: type.id,
          },
          ctx,
        );

        return result;
      },
    }),
  }),
});
