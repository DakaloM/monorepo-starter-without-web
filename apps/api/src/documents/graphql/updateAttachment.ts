import { builder } from '~/graphql/builder';

import { isActiveUser } from '~/account/shield';
import { getConfidentialAssignees } from '~/assignee/query';
import { EventAction } from '~/event';
import { TypeIdentifier } from '~/type';

import { Attachment } from '../attachment';
import { AttachmentSchema } from './attachment';

const UpdateAttachmentInput = builder.inputType('UpdateAttachmentInput', {
  fields: (t) => ({
    id: t.string({ required: true }),
    confidential: t.boolean({ required: false }),
    assignees: t.stringList({ required: false }),
    description: t.string({ required: false }),
    tags: t.stringList({ required: false }),
  }),
  description: 'Input type for uploading an attachment.',
});

builder.mutationField('updateAttachment', (t) =>
  t.field({
    args: {
      input: t.arg({ type: UpdateAttachmentInput, required: true }),
    },
    type: AttachmentSchema,
    shield: isActiveUser,
    description: 'Updates an attachment.',
    resolve: async (_parent, { input: args }, ctx) => {
      return Attachment.transaction(ctx.db, async (trx) => {
        const attachment = await Attachment.query(trx).patchAndFetchById(args.id, {
          confidential: args.confidential || undefined,
          description: args.description || undefined,
        });

        const assignees = getConfidentialAssignees(
          {
            assignees: args.assignees || [],
            confidential: args.confidential || false,
          },
          ctx.auth?.user?.id as string,
        );

        await ctx.createEvent<Attachment>(
          {
            action: EventAction.Update,
            refId: attachment.id,
            refTypeIdentifier: TypeIdentifier.Attachment,
            change: args as Partial<Attachment>,
            description: attachment.name,
            confidential: false,
          },
          trx,
        );

        return attachment;
      });
    },
  }),
);
