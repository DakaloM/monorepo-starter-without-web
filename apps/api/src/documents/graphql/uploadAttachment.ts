import { builder } from '~/graphql/builder';

import { isActiveUser } from '~/account/shield';
import { createAttachment } from '~/documents/mutations';
import { TypeIdentifier } from '~/type';

import { AttachmentSchema } from './attachment';

const UploadAttachmentInput = builder.inputType('UploadAttachmentInput', {
  fields: (t) => ({
    name: t.string({ required: true }),
    contentType: t.string({ required: true }),
    contentLength: t.int({ required: true }),
    confidential: t.boolean({ required: true }),
    description: t.string({ required: false }),
    refId: t.string({ required: true }),
    refTypeIdentifier: t.string({ required: true }),
    assignees: t.stringList({ required: true }),
  }),
  description: 'Input type for uploading an attachment.',
});

builder.mutationField('uploadAttachment', (t) =>
  t.field({
    args: {
      input: t.arg({ type: UploadAttachmentInput, required: true }),
    },
    type: AttachmentSchema,
    shield: isActiveUser,
    description: 'Uploads an attachment.',
    resolve: async (_parent, { input: args }, ctx) => {
      const attachment = await createAttachment(
        {
          contentLength: args.contentLength,
          contentType: args.contentType,
          name: args.name,
          refId: args.refId,
          refTypeIdentifier: args.refTypeIdentifier as TypeIdentifier,
          confidential: args.confidential,
          assignees: args.assignees,
        },
        ctx,
      );

      return attachment;
    },
  }),
);
