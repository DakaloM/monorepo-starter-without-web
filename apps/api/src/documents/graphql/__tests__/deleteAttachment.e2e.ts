import { User } from '~/account/user';
import { Case } from '~/case';
import { Category, CategoryName } from '~/case/category';
import { Context } from '~/context';
import { Attachment, AttachmentState } from '~/documents/attachment';
import { createContext, destroyContext } from '~/test/context';
import { createEnv, destroyEnv, Environment } from '~/test/environment';
import { createFactory, Factory } from '~/test/factory';
import { gql, request } from '~/test/graphql';
import { Type, TypeIdentifier } from '~/type';

describe('documents/graphql/deleteAttachment', () => {
  let env: Environment;
  let ctx: Context;
  let factory: Factory;
  let type: Type;
  let numCase: Case;
  let category: Category;
  const query = gql`
    mutation DeleteAttachment($id: ID!) {
      deleteAttachment(id: $id) {
        id
      }
    }
  `;

  beforeAll(async () => {
    env = await createEnv({ db: true, cache: true, server: true });
  });

  afterAll(async () => {
    await destroyEnv(env);
  });

  let user: User;

  let attachment: Attachment;

  beforeEach(async () => {
    ctx = await createContext(env);
    factory = createFactory(ctx);
    const region = await factory.insert('region');
    const tenant = await factory.insert('tenant', { regionId: region.id });
    user = await factory.insert('user', {tenantId: tenant.id});
    const types = await factory.seed('types');
    category = await factory.insert('category');
    numCase = await factory.insert('case', { categoryId: category.id });
    type = types[TypeIdentifier.Case];

    attachment = await factory.insert('attachment', {
      creatorId: user.id,
      refId: numCase.id,
      refTypeId: type.id,
    });

    ctx.auth.update(user);
  });

  afterEach(async () => {
    await destroyContext(ctx);
  });

  it('should add deletedAt to attachment', async () => {
    const { data } = await request(env)
      .withContext(ctx)
      .query(query)
      .variables({ id: attachment.id })
      .expectNoErrors();

    expect(data.deleteAttachment.id).toEqual(attachment.id);

    const deletedAttachment = await Attachment.query(ctx.db)
      .findById(attachment.id)
      .throwIfNotFound();

    expect(deletedAttachment.deletedAt).toBeDefined();
  });

  it('should not change deletedAt when already deleted', async () => {
    const deletedAt = new Date('2020-01-01');
    const attachment = await factory.insert('attachment', {
      creatorId: user.id,
      refId: numCase.id,
      refTypeId: type.id,
      sequence: 2,
      deletedAt,
    });

    const { data } = await request(env)
      .withContext(ctx)
      .query(query)
      .variables({ id: attachment.id })
      .expectNoErrors();

    expect(data.deleteAttachment.id).toEqual(attachment.id);

    const deletedAttachment = await Attachment.query(ctx.db)
      .findById(attachment.id)
      .throwIfNotFound();

    expect(deletedAttachment.deletedAt).toEqual(deletedAt);
    expect(deletedAttachment.deletedAt).toMatchSnapshot();
  });
});
