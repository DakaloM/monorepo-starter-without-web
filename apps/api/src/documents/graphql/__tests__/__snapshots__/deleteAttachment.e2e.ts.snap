// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`documents/graphql/deleteAttachment > should not change deletedAt when already deleted 1`] = `2020-01-01T00:00:00.000Z`;
